src/locale/keymap-util.c

```diff
@@ -780,3 +782,141 @@ int x11_convert_to_vconsole(Context *c) {
 
         return modified;
 }
+
+static
+bool locale_encoding_is_utf8_or_unspecified(const char *locale) {
+        return !strstr(locale, ".") || endswith_no_case(locale, "UTF-8");
+}
+
+bool locale_gen_check_available(void) {
+        if (access("/usr/sbin/locale-gen", X_OK) < 0) {
+                if (errno != ENOENT)
+                        log_warning_errno(errno, "Unable to determine whether /usr/sbin/locale-gen exists: %m");
+                return false;
+        }
+        if (access("/etc/locale.gen", F_OK) < 0) {
+                if (errno != ENOENT)
+                        log_warning_errno(errno, "Unable to determine whether /etc/locale.gen exists: %m");
+                return false;
+        }
+        return true;
+}
+
+int locale_gen_enable_locale(char *locale[_VARIABLE_LC_MAX]) {
+        _cleanup_fclose_ FILE *fr = NULL, *fw = NULL;
+        _cleanup_(unlink_and_freep) char *temp_path = NULL;
+        _cleanup_free_ char *locale_entry = NULL;
+        bool locale_enabled = false, first_line = false;
+        struct stat st;
+        int r;
+
+        if (isempty(locale[VARIABLE_LANG]))
+                return -EINVAL;
+
+        if (locale_encoding_is_utf8_or_unspecified(locale[VARIABLE_LANG])) {
+                locale_entry = strjoin(locale[VARIABLE_LANG], " UTF-8");
+                if (!locale_entry)
+                        return -ENOMEM;
+        } else
+                return -EINVAL; /* We do not process non-UTF-8 locale */
+
+        fr = fopen("/etc/locale.gen", "re");
+        if (!fr)
+                return -errno;
+
+        r = fopen_temporary("/etc/locale.gen", &fw, &temp_path);
+        if (r < 0)
+                return r;
+
+        /* apply mode of the original file to new file */
+        if (stat("/etc/locale.gen", &st) < 0)
```

Comment:
please use `fstat(fileno(fr), &st)` here, so that we read the stat data from the actually opened file (feels safer, given that the file might be replaced any time)